R/add_weights.R

```diff
@@ -0,0 +1,70 @@
+#' Add a weight variable using the sample frame
+#'
+#' @param .dataset the clean dataframe
+#' @param sample_data sample dataframe including poplution numbers and the strata
+#' @param strata_column_dataset name of strata column in the clean dataframe
+#' @param strata_column_sample name of strata column in the sample dataframe
+#' @param population_column name of population column in the sample dataframe
+#' @param weight_column name of the added weight column. By default "weight"
+#'
+#' @return The clean dataset with 1 new column: weight
+#' @export
+#'
+#' @examples
+#' clean_data <- data.frame(uuid = c(1,2,3,4,5,6,7,8),
+#'                          strata = c("strata1","strata2","strata1",
+#'                                     "strata2","strata1","strata2",
+#'                                     "strata1","strata1"))
+#' sample <- data.frame(strata = c("strata1","strata2"),
+#'                      population = c(30000,50000))
+#'
+#' clean_data_weighted <- clean_data %>%
+#'     add_weights(sample,
+#'                 strata_column_dataset = "strata",
+#'                 strata_column_sample = "strata",
+#'                 population_column = "population")
+add_weights<- function(.dataset,
+                        sample_data,
+                        strata_column_dataset = NULL,
+                        strata_column_sample = NULL,
+                        population_column = NULL,
+                        weight_column = "weight"){
+
+  # make dataset a dataframe
+  .dataset <- as.data.frame(.dataset)
+
+  # If strata_column do not exist in sample_data or dataset
+  if(!strata_column_sample %in% names(sample_data))
+    stop("Cannot find the defined strata column in the provided sample frame.")
+  if(!strata_column_dataset %in% names(.dataset))
+    stop("Cannot find the defined strata column in the provided dataset.")
+
```

Comment:
Perhaps we can add -

1. Error message if all the strata from the dataset are not available in the sample frame 
2. Warning message -if all the strata from the sample frame are not available in the dataset [I am suggesting warning, because sometimes its normal to have population data across the country, including the areas where we are having access issue]

Reply:
- if there is a missing strata in the dataset. the weights df will return only NA.
- if you add a na.rm = T, the weights will including the population of the missing strata, hence not correct. 
- I am not really sure what Mehedi wanted to achieve here. We can change the warning to error until he comes back.

#missing strata in dataset : does not run
set.seed(2133)
my_data <- data.frame(aa = runif(100),
                      strata = sample(LETTERS[1:5],size = 100, T))
my_sf <- data.frame(strata = LETTERS[1:6], 
                    pop = c(10000,10000,20000,30000,5000,5000))
add_weights(my_data,
            my_sf,
            strata_column_dataset = "strata",
            strata_column_sample = "strata",
            population_column = "pop") %>% 
  dplyr::summarise(sum(weight))

#missing strata in sampling frame : ok  
set.seed(2133)
my_data <- data.frame(aa = runif(100),
                      strata = sample(LETTERS[1:6],size = 100, T))
my_sf <- data.frame(strata = LETTERS[1:5], 
                    pop = c(10000,10000,20000,30000,5000))
add_weights(my_data,
            my_sf,
            strata_column_dataset = "strata",
            strata_column_sample = "strata",
            population_column = "pop") %>% 
  dplyr::summarise(sum(weight))

Reply:
For the third point, I think what Mehedi meant is that the sampling frame was done, then data collection happened in fewer locations than what was decided because of field reasons. I understand that this then requires either the change of the whole sample frame document to match the collected data, as filtering out the strata not collected in the sample frame will not resolve the issue.
I changed it for the time being to error as mentioned, but we can look again at it.